File: Composite/Composite.cpp
#include <iostream>
#include <vector>

using namespace std;

// Component interface
class Employee {
public:
    virtual void showDetails() = 0;
};

// Leaf class
class Developer : public Employee {
private:
    string name;

public:
    Developer(const string& name) : name(name) {}

    void showDetails() override {
        cout << "Developer: " << name << endl;
    }
};

// Composite class
class Manager : public Employee {
private:
    string name;
    vector<Employee*> subordinates;

public:
    Manager(const string& name) : name(name) {}

    void addSubordinate(Employee* employee) {
        subordinates.push_back(employee);
    }

    void showDetails() override {
        cout << "Manager: " << name << endl;
        for (Employee* subordinate : subordinates) {
            subordinate->showDetails();
        }
    }
};

int main() {
    Manager ceo("John (CEO)");
    Manager engineeringHead("Alice (Engineering Head)");
    Developer dev1("Bob");
    Developer dev2("Charlie");

    engineeringHead.addSubordinate(&dev1);
    engineeringHead.addSubordinate(&dev2);
    ceo.addSubordinate(&engineeringHead);

    ceo.showDetails();

    return 0;
}

File: Composite/Composite.java
import java.util.ArrayList;
import java.util.List;

// Component interface
interface Employee {
    void showDetails();
}

// Leaf class
class Developer implements Employee {
    private String name;

    public Developer(String name) {
        this.name = name;
    }

    public void showDetails() {
        System.out.println("Developer: " + name);
    }
}

// Composite class
class Manager implements Employee {
    private String name;
    private List<Employee> subordinates = new ArrayList<>();

    public Manager(String name) {
        this.name = name;
    }

    public void addSubordinate(Employee employee) {
        subordinates.add(employee);
    }

    public void showDetails() {
        System.out.println("Manager: " + name);
        for (Employee subordinate : subordinates) {
            subordinate.showDetails();
        }
    }
    
        public static void main(String[] args) {
        Manager ceo = new Manager("John (CEO)");
        Manager engineeringHead = new Manager("Alice (Engineering Head)");
        Developer dev1 = new Developer("Bob");
        Developer dev2 = new Developer("Charlie");

        engineeringHead.addSubordinate(dev1);
        engineeringHead.addSubordinate(dev2);
        ceo.addSubordinate(engineeringHead);

        ceo.showDetails();
    }
}
